File: NodeDemo/src/services/externalApiService.ts
/**
 * Simulates external API calls with realistic delays
 */

// Helper to create random delay
const randomDelay = (min: number, max: number): number => {
    return Math.floor(Math.random() * (max - min + 1)) + min;
};

/**
 * Helper to create a promise that resolves after a delay
 */
const delay = (ms: number): Promise<void> => {
    return new Promise(resolve => setTimeout(resolve, ms));
};

export class ExternalApiService {
    /**
     * ASYNC: Simulates calling an external API asynchronously
     * Event loop is free to handle other requests during the network wait
     */
    async getWeatherAsync(city: string): Promise<string> {
        // Simulate network latency (100-300ms)
        await delay(randomDelay(100, 300));

        return `Weather in ${city}: Sunny, 72°F`;
    }

    /**
     * BLOCKING (BAD): Simulates a blocking external API call
     * Blocks the event loop - catastrophic in Node.js!
     */
    getWeatherBlocking(city: string): string {
        // BAD: Blocks the entire event loop
        const delayMs = randomDelay(100, 300);
        const start = Date.now();
        while (Date.now() - start < delayMs) {
            // Busy wait - disaster for Node.js!
        }

        return `Weather in ${city}: Sunny, 72°F`;
    }

    /**
     * ASYNC: Simulates calling multiple external APIs in parallel
     */
    async getEnrichedDataAsync(userId: string): Promise<Record<string, string>> {
        // All API calls run in parallel
        const [profile, preferences, analytics] = await Promise.all([
            delay(randomDelay(80, 150)).then(() => `Profile data for ${userId}`),
            delay(randomDelay(60, 120)).then(() => 'Dark mode enabled'),
            delay(randomDelay(90, 180)).then(() => 'Last login: 2 hours ago')
        ]);

        return {
            profile,
            preferences,
            analytics
        };
    }

    /**
     * BLOCKING (BAD): Simulates calling multiple external APIs sequentially
     * Each call blocks the event loop
     */
    getEnrichedDataBlocking(userId: string): Record<string, string> {
        // BAD: Sequential blocking calls
        let start = Date.now();
        let delayMs = randomDelay(80, 150);
        while (Date.now() - start < delayMs) { /* busy wait */ }
        const profile = `Profile data for ${userId}`;

        start = Date.now();
        delayMs = randomDelay(60, 120);
        while (Date.now() - start < delayMs) { /* busy wait */ }
        const preferences = 'Dark mode enabled';

        start = Date.now();
        delayMs = randomDelay(90, 180);
        while (Date.now() - start < delayMs) { /* busy wait */ }
        const analytics = 'Last login: 2 hours ago';

        return {
            profile,
            preferences,
            analytics
        };
    }
}

File: CSharpDemo/AsyncAwaitWebApi/Services/ExternalApiService.cs
namespace AsyncAwaitWebApi.Services;

/// <summary>
/// Simulates external API calls with realistic delays
/// </summary>
public class ExternalApiService
{
    private readonly Random _random = new Random();

    /// <summary>
    /// ASYNC: Simulates calling an external API asynchronously
    /// </summary>
    public async Task<string> GetWeatherAsync(string city)
    {
        // Simulate network latency for external API call (100-300ms)
        await Task.Delay(_random.Next(100, 300));

        return $"Weather in {city}: Sunny, 72°F";
    }

    /// <summary>
    /// SYNC: Simulates calling an external API synchronously (BLOCKING)
    /// </summary>
    public string GetWeatherSync(string city)
    {
        // BAD: Blocking the thread for network I/O
        Thread.Sleep(_random.Next(100, 300));

        return $"Weather in {city}: Sunny, 72°F";
    }

    /// <summary>
    /// ASYNC: Simulates calling multiple external APIs
    /// </summary>
    public async Task<Dictionary<string, string>> GetEnrichedDataAsync(string userId)
    {
        // Simulate calling 3 different external services in parallel
        var profileTask = Task.Delay(_random.Next(80, 150))
            .ContinueWith(_ => $"Profile data for {userId}");

        var preferencesTask = Task.Delay(_random.Next(60, 120))
            .ContinueWith(_ => "Dark mode enabled");

        var analyticsTask = Task.Delay(_random.Next(90, 180))
            .ContinueWith(_ => "Last login: 2 hours ago");

        await Task.WhenAll(profileTask, preferencesTask, analyticsTask);

        return new Dictionary<string, string>
        {
            ["profile"] = profileTask.Result,
            ["preferences"] = preferencesTask.Result,
            ["analytics"] = analyticsTask.Result
        };
    }

    /// <summary>
    /// SYNC: Simulates calling multiple external APIs sequentially (BAD)
    /// </summary>
    public Dictionary<string, string> GetEnrichedDataSync(string userId)
    {
        // BAD: Sequential blocking calls - much slower than parallel
        Thread.Sleep(_random.Next(80, 150));
        var profile = $"Profile data for {userId}";

        Thread.Sleep(_random.Next(60, 120));
        var preferences = "Dark mode enabled";

        Thread.Sleep(_random.Next(90, 180));
        var analytics = "Last login: 2 hours ago";

        return new Dictionary<string, string>
        {
            ["profile"] = profile,
            ["preferences"] = preferences,
            ["analytics"] = analytics
        };
    }
}
